File: 0912/solution.java
import java.util.Random;

class Solution {
    public int[] sortArray(int[] nums) {
        quickSort(nums, 0, nums.length - 1);
        return nums;
    }

    private void quickSort(int[] nums, int l, int r) {
        if (l >= r) {
            return;
        }
        Random rand = new Random();
        int p = rand.nextInt(r - l + 1) + l;
        int pivot = nums[p];
        nums[p] = nums[r];
        nums[r] = pivot;
        int i = l;
        int j = l;
        int k = r;

        while (j <= k) {
            if (nums[j] < pivot) {
                int temp = nums[j];
                nums[j] = nums[i];
                nums[i] = temp;
                i++;
                j++;
            } else if (nums[j] > pivot) {
                int temp = nums[j];
                nums[j] = nums[k];
                nums[k] = temp;
                k--;
            } else {
                j++;
            }
        }
        quickSort(nums, l, i - 1);
        quickSort(nums, k + 1, r);
    }
}

// time O(nlogn)
// space O(logn)

File: 0912/solution.py
import random


class Solution:
    def sortArray(self, nums: list[int]) -> list[int]:
        def quicksort(l, r):
            if l >= r:
                return

            p = random.randint(l, r)
            nums[r], nums[p] = nums[p], nums[r]
            pivot = nums[r]

            i, j, k = l, l, r
            while j <= k:
                if nums[j] < pivot:
                    nums[i], nums[j] = nums[j], nums[i]
                    i += 1
                    j += 1
                elif nums[j] > pivot:
                    nums[j], nums[k] = nums[k], nums[j]
                    k -= 1
                else:
                    j += 1

            quicksort(l, i - 1)
            quicksort(k + 1, r)

        quicksort(0, len(nums) - 1)
        return nums
